src/Tools/ExternalAccess/Razor/RazorAnalyzerAssemblyResolver.cs

```diff
@@ -18,41 +19,31 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
     [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
     internal class RazorAnalyzerAssemblyResolver() : IAnalyzerAssemblyResolver
     {
-        private static Func<AssemblyName, Assembly?>? s_assemblyResolver;
+        private const string RazorCompilerAssemblyName = "Microsoft.CodeAnalysis.Razor.Compiler";
+        private const string RazorUtilsAssemblyName = "Microsoft.AspNetCore.Razor.Utilities.Shared";
+        private const string ObjectPoolAssemblyName = "Microsoft.Extensions.ObjectPool";
 
-        private static readonly HashSet<AssemblyName> s_assembliesRequested = [];
+        private static readonly object s_loaderLock = new();
 
-        private static readonly object s_resolverLock = new();
-
-        /// <summary>
-        /// Attempts to set the assembly resolver. Will only succeed if the <paramref name="canaryAssembly"/> has not already been requested to load and the resolver has not been set previously.
-        /// </summary>
-        /// <param name="resolver">The resolver function to set.</param>
-        /// <param name="canaryAssembly">The name of an assembly that is checked to see if it has been requested to load. Setting the resolver will fail if this has already been requested.</param>
-        /// <returns><c>true</c> when the resolver was successfully set.</returns>
-        internal static bool TrySetAssemblyResolver(Func<AssemblyName, Assembly?> resolver, AssemblyName canaryAssembly)
+        public static Assembly? ResolveRazorAssembly(AssemblyName assemblyName, string directoryName)
         {
-            lock (s_resolverLock)
+            if (assemblyName.Name is RazorCompilerAssemblyName or RazorUtilsAssemblyName or ObjectPoolAssemblyName)
             {
-                if (s_assemblyResolver is null && !s_assembliesRequested.Contains(canaryAssembly))
+                var assembly = Path.Combine(directoryName, $"{assemblyName.Name}.dll");
+                lock (s_loaderLock)
                 {
-                    s_assemblyResolver = resolver;
-                    return true;
+                    var compilerContext = AssemblyLoadContext.GetLoadContext(typeof(Compilation).Assembly)!;
+                    if (compilerContext.Assemblies.SingleOrDefault(a => a.GetName().Name == assemblyName.Name) is Assembly loadedAssembly)
+                    {
+                        return loadedAssembly;
+                    }
+                    return compilerContext.LoadFromAssemblyPath(assembly);
```

Comment:
Note that here we're loading razor into the 'main' compiler ALC. We could create our own ALC to load into, but it will need to be custom code that defers to the main ALC for things like MS.CA.dll, that seemed overkill so I just used the main one (which is already conceptually the 'shared' one anyway).